core: Fix highlight nick config save/load

digitalcircuit opened the pull request:
## In short

* Modernize `CoreHighlightRuleManager` via `SyncableObject` functions, following `DccConfig`
  * Fixes highlight nick configuration being reset on core/monolithic restart
  * Simplifies code
  * Fixes regression from [highlight protocol pull request #370](https://github.com/quassel/quassel/pull/370 )
* Save remote highlight rules if changed before importing local
  * Fixes losing unsaved rules when importing

Criteria | Rank | Reason
---------|---------|-------------
Impact | ★★★ *3/3* | User-facing common functionality - remembering highlight nick options
Risk | ★★☆ *2/3* | Core-side change, revertible with configuration loss
Intrusiveness | ★☆☆ *1/3* | Small set of changes, shouldn't interfere with other pull requests

## Rationale
Persistence for **`Highlight Nicks`** is needed, for otherwise restarting the core **or the monolithic client** will reset nickname highlighting options, re-enabling if disabled and limiting to current nick if more had been chosen.

### Breaking changes
Anyone currently using the `master` branch will have to re-set-up their core-side highlight rules one more time (*sorry*).  Afterwards, settings persist as expected.

Upgrading from `0.12.5` or earlier works fine.

## Testing
### Steps
1.  Set up core and client
2.  In `Configure Quassel…` → `Remote Highlights`, change **`Highlight Nicks`** to something non-default
    * E.g. `All Nicks from Identity`, `Case sensitive`
3.  Apply settings, shutdown core
4.  Launch core, reconnect
5.  Open configuration to `Remote Highlights` page

### Before
**`Highlight Nicks`** settings reset to default on every load.

### After
Even after core and client restarts, **`Highlight Nicks`** settings persist and have effect on core highlight behavior.

digitalcircuit commented:
I did try the method from `CoreDccConfig`, of `toVariantMap()` and `update()`.  On upgrade, it didn't apply the right default settings (*picked `None` instead of `Current nick` despite header having initialized values*).  On a blank/fresh configuration, the core crashed.

I almost certainly was doing it wrong, just probably comes down to time to figure things out.

Sput42 commented on src/core/corehighlightrulemanager.cpp:
```diff
@@ -34,24 +37,49 @@ CoreHighlightRuleManager::CoreHighlightRuleManager(CoreSession *parent)
+    QVariantMap highlightConfig;
+
+    highlightConfig["highlightNick"] = highlightNick();
+    highlightConfig["nicksCaseSensitive"] = nicksCaseSensitive();
```
I would appreciate keys in maps starting with capital letter. I know we're not consistent everywhere (CoreInfo comes to mind), but where we can change it without breaking compat, or when introducing new maps, we should do it right.

Sput42 replied:
I retract this request, I now understand that you plan to initialize the properties from the map using fromVariantMap(). In this case, the keys need to match the property names exactly, so lowerCamelCase is ok here.

Sput42 commented on src/core/corehighlightrulemanager.cpp:
```diff
@@ -34,24 +37,49 @@ CoreHighlightRuleManager::CoreHighlightRuleManager(CoreSession *parent)
+
+
+void CoreHighlightRuleManager::localInitSetExtraConfig(const QVariantMap &highlightConfig) {
+    // TODO: Clean this up in a way that properly handles settings
```
I'm not quite sure I understand the comment...

Sput42 commented on src/core/corehighlightrulemanager.h:
```diff
@@ -50,5 +50,21 @@ public slots:
+     *
+     * @return Additional highlight configuration as a variant map
+     */
+    QVariantMap localInitExtraConfig();
```
Both these methods are very confusingly named IMHO. The init* and initSet* methods have a very specific purpose in the context of syncable objects, but for "local" setters we should just be using normal names such as setHighlightConfig() or whatever. Maybe I'm not understanding correctly what you plan to do in the final version, though?

Sput42 commented on src/core/corehighlightrulemanager.cpp:
```diff
@@ -34,24 +37,49 @@ CoreHighlightRuleManager::CoreHighlightRuleManager(CoreSession *parent)
 }
+
+
+void CoreHighlightRuleManager::localInitSetExtraConfig(const QVariantMap &highlightConfig) {
```
Linebreak

Sput42 commented on src/core/corehighlightrulemanager.cpp:
```diff
@@ -34,24 +37,49 @@ CoreHighlightRuleManager::CoreHighlightRuleManager(CoreSession *parent)
+    // Make sure the default for _highlightNick is "CurrentNick" if not set
+    setHighlightNick(HighlightNickType(
+                         highlightConfig.value("highlightNick",
+                                               HighlightNickType::CurrentNick).toInt()));
```
Is this weirdly indented, or does the Github UI fool me?

Sput42 commented on src/core/corehighlightrulemanager.cpp:
```diff
@@ -34,24 +37,49 @@ CoreHighlightRuleManager::CoreHighlightRuleManager(CoreSession *parent)
+}
+
+
+QVariantMap CoreHighlightRuleManager::localInitExtraConfig() {
```
Linebreak

digitalcircuit commented:
Thank you for the comments and review!

As discussed on IRC, this was more of a hackish "get something working for a beta deadline" approach in case `update()` for loading/`toVariantMap()` for saving can't be used [as it is in `CoreDccConfig`](https://github.com/quassel/quassel/blob/master/src/core/coredccconfig.cpp ).  My initial efforts looking into that resulted in crashes, but I'll try again.

As to why?  It'd be a lot cleaner to get rid of the manual `initSet`/`init` calls since the `update`/`toVariantMap` should automatically call them *and* automatically include all `Q_PROPERTY` items.  I'll have to test upgrading and clean install configuration to make sure the right default settings are being used.

I'll keep in mind the corrections you provided where applicable, and should we need a separate save/load method for nick configuration after all, I'll certainly apply the rest of the corrections to this, too.

digitalcircuit commented:
I've reworked this with the `update()`/`toVariantMap()` options, and everything on Linux seems to work!

In all cases below, I started/set up core, connected client, configured, shut down core, then started up core, checked settings persisted.

* Fresh configuration: **pass**
* Upgrade from faux-`0.12.5`: **pass**
  * I just deleted `HighlightRuleList` from the `user_setting` table on a `0.13` build
* Upgrade from `master`: **pass**
  * All existing core-side highlights are deleted, just this once

More extensive testing could be done, but this seems to at least make the situation better in the basic tests I used.
